core/examples/ts_module_loader.rs

```diff
@@ -0,0 +1,119 @@
+//! It will only transpile, not typecheck (like Deno's `--no-check` flag).
+
+use std::path::PathBuf;
+use std::pin::Pin;
+use std::rc::Rc;
+
+use anyhow::anyhow;
+use anyhow::bail;
+use anyhow::Error;
+use deno_ast::MediaType;
+use deno_ast::ParseParams;
+use deno_ast::SourceTextInfo;
+use deno_core::resolve_import;
+use deno_core::resolve_path;
+use deno_core::JsRuntime;
+use deno_core::ModuleLoader;
+use deno_core::ModuleSource;
+use deno_core::ModuleSourceFuture;
+use deno_core::ModuleSpecifier;
+use deno_core::ModuleType;
+use deno_core::RuntimeOptions;
+use futures::FutureExt;
+
+struct TypescriptModuleLoader;
+
+impl ModuleLoader for TypescriptModuleLoader {
+  fn resolve(
+    &self,
+    specifier: &str,
+    referrer: &str,
+    _is_main: bool,
+  ) -> Result<ModuleSpecifier, Error> {
+    Ok(resolve_import(specifier, referrer)?)
+  }
+
+  fn load(
+    &self,
+    module_specifier: &ModuleSpecifier,
+    _maybe_referrer: Option<ModuleSpecifier>,
+    _is_dyn_import: bool,
+  ) -> Pin<Box<ModuleSourceFuture>> {
+    let module_specifier = module_specifier.clone();
+    async move {
+      let path = module_specifier
+        .to_file_path()
+        .map_err(|_| anyhow!("Only file: URLs are supported."))?;
+
+      let media_type = MediaType::from(&path);
+      let (module_type, do_transpile) = match MediaType::from(&path) {
+        MediaType::JavaScript | MediaType::Mjs | MediaType::Cjs => {
+          (ModuleType::JavaScript, false)
+        }
+        MediaType::Jsx => (ModuleType::JavaScript, true),
+        MediaType::TypeScript
+        | MediaType::Mts
+        | MediaType::Cts
+        | MediaType::Dts
+        | MediaType::Dmts
+        | MediaType::Dcts
+        | MediaType::Tsx => (ModuleType::JavaScript, true),
+        MediaType::Json => (ModuleType::Json, false),
+        _ => bail!("Unknown extension {:?}", path.extension()),
+      };
+
+      let code = std::fs::read_to_string(&path)?;
+      let code = if do_transpile {
+        let parsed = deno_ast::parse_module(ParseParams {
+          specifier: module_specifier.to_string(),
+          source: SourceTextInfo::from_string(code),
+          media_type,
+          capture_tokens: false,
+          scope_analysis: false,
+          maybe_syntax: None,
+        })?;
+        parsed.transpile(&Default::default())?.text
+      } else {
+        code
+      };
+      let module = ModuleSource {
+        code,
+        module_type,
+        module_url_specified: module_specifier.to_string(),
+        module_url_found: module_specifier.to_string(),
+      };
+      Ok(module)
+    }
+    .boxed_local()
+  }
+}
+
+fn main() -> Result<(), Error> {
+  let mut js_runtime = JsRuntime::new(RuntimeOptions {
+    module_loader: Some(Rc::new(TypescriptModuleLoader)),
+    ..Default::default()
+  });
+
+  let main_module = {
+    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+    path.push("examples/ts_module_loader.ts");
+    resolve_path(path.to_str().unwrap())?
+  };
```

Comment:
Could you align this example to `core/examples/fs_module_loader.rs` to allow executing arbitrary TS file specified as a CLI arg?

Reply:
Done.